src/main/java/org/jboss/msc/service/ServiceControllerImpl.java

```diff
@@ -1621,9 +1593,11 @@ final boolean execute() {
     }
 
     private abstract class DependentsControllerTask extends ControllerTask {
-        protected final Dependent[][] dependents;
+        protected final Map<ServiceName, Dependent[]> dependents;
+        protected final Dependent[] children;
         private DependentsControllerTask() {
-            dependents = getDependents();
+            dependents = getDependentsByDependencyName();
+            children = ServiceControllerImpl.this.children.toScatteredArray(NO_DEPENDENTS);
```

Comment:
If you want to have a look in advance this is the change I am talking about:
https://github.com/ropalka/jboss-msc/commit/0bfdddf8be98b5c5332ad076019fe54c8647e8de#diff-fd8643d01445d7e64971875e1b6a1d26R1684